bluefog/common/operations.cc

```diff
@@ -103,54 +103,88 @@ void BackgroundThreadLoop(BluefogGlobalState& state) {
   mpi_context.Finalize(mpi_ctx_manager);
 
 #if HAVE_NCCL
-  nccl_context.Finalize();
+  if (nccl_context.is_initialized) {
+    nccl_context.Finalize();
+  }
 #endif
 }
 
-int DetermineController(const TensorTableEntry& entry) {
-  if (entry.mpi_ops_type != MPIOpsType::ALLREDUCE &&
-      entry.mpi_ops_type != MPIOpsType::BROADCAST &&
-      entry.mpi_ops_type != MPIOpsType::ALLGATHER)
-    return 0;
-  bool have_nccl = false;
+Vendor DetermineController(const TensorTableEntry& entry) {
+  bool nccl_impl_available = false;
+  bool force_mpi = false;
+  bool built_with_nccl = false;
 #if HAVE_NCCL
-  have_nccl = true;
+  built_with_nccl = true;
 #endif
-  return (have_nccl && entry.device != CPU_DEVICE_ID) ? 1 : 0;
+  char* by_mpi_env;
+  switch (entry.mpi_ops_type) {
+    case MPIOpsType::ALLREDUCE:
+      by_mpi_env = std::getenv("BLUEFOG_ALLREDUCE_BY_MPI");
+      force_mpi = (by_mpi_env != nullptr) && (*by_mpi_env == '1');
+      nccl_impl_available = true;
+      break;
+    case MPIOpsType::BROADCAST:
+      by_mpi_env = std::getenv("BLUEFOG_BROADCAST_BY_MPI");
+      force_mpi = (by_mpi_env != nullptr) && (*by_mpi_env == '1');
+      nccl_impl_available = true;
+      break;
+    case MPIOpsType::ALLGATHER:
+      by_mpi_env = std::getenv("BLUEFOG_ALLGATHER_BY_MPI");
+      force_mpi = (by_mpi_env != nullptr) && (*by_mpi_env == '1');
+      nccl_impl_available = true;
+      break;
+    case MPIOpsType::NEIGHBOR_ALLGATHER:
+      by_mpi_env = std::getenv("BLUEFOG_NEIGHBOR_ALLGATHER_BY_MPI");
+      force_mpi = (by_mpi_env != nullptr) && (*by_mpi_env == '1');
+      nccl_impl_available = true;
+      break;
+    case MPIOpsType::NEIGHBOR_ALLREDUCE:
+      by_mpi_env = std::getenv("BLUEFOG_NEIGHBOR_ALLREDUCE_BY_MPI");
+      force_mpi = (by_mpi_env != nullptr) && (*by_mpi_env == '1');
+      nccl_impl_available = true;
+      break;
+    default:
+      nccl_impl_available = false;
+  }
+
+  if (!built_with_nccl || !nccl_impl_available || force_mpi) return Vendor::MPI;
+  return entry.device != CPU_DEVICE_ID ? Vendor::NCCL : Vendor::MPI;
```

Comment:
Why NCCL ensures entry.device != CPU_DEVICE_ID?

Reply:
Because NCCL controller won't work for CPU cases.